### PR TITLE
Namespace for

### PR DESCRIPTION
**Pull request by privat**
Better implementation for `full_name` and `c_name` that make them shorter and use the project as the namespace of public entities.

Examples of fullnames of entities used in messages

* A public class `A` defined in the module `m` of the project `p` was `p::m::A`, now it is `p::A`.
* A public method `x` defined in a class `A` in the module `m` of the project `p` was `p::m::A::x`, now it is `p::A::x`

Exeaple of fullnames for entities that are used internally:

* The refinement of `A` in a module `p::n` was `p::n#p::m::A`, now it is `p::n#A`
* The redefinition of `x` in a class `B` in `p::n` was `p::n#p::m::B#p::m::A::x`, now it is `p::n#B#A::x`

The `c_name` get some comparable simplifications, so C identifiers will be simpler and will less overflow in unwind or in valgrind.

Note: please ignore the first commits that come from #1069

### CODE REVIEW

File: src/model/mmodule.nit [hunk 1]
```diff
@@ -104,6 +112,22 @@ class MModule
...
 	end
 
+	# The namespace used for public entities.
+	#
+	# Public entities use only the project as a namespace.
+	# Private entities use the `full_name` (i.e. "project::module")
+	#
+	# This method is used by entities to implement their `full_name`.
+	fun namespace_for(v: MVisibility): String do
+		if v <= private_visibility then return full_name
```
Comment (Morriar) [inline src/model/mmodule.nit]: By reading the synopsis one can think that this method do a noop for private entities.

File: src/model/mmodule.nit [hunk 2]
```diff
@@ -117,6 +141,19 @@ class MModule
 		return res
 	end
 
+	# C identifier version of `namespace_for`.
+	# See `c_name`
+	#
+	# This method is used by entities to implement their `c_name`.
+	fun c_namespace_for(v: MVisibility): String do
+		if v <= private_visibility then return c_name
```
Comment (Morriar) [inline src/model/mmodule.nit]: This should be stated in the comment.

File: src/model/model.nit [hunk 3]
```diff
@@ -539,17 +543,29 @@ class MClassDef
...
 	end
 
 	redef var c_name is lazy do
 		if is_intro then
 			return mclass.c_name
+		else if mclass.intro_mmodule.mproject == mmodule.mproject and mclass.visibility > private_visibility then
+			return "{mmodule.c_name}_0_{mclass.name.to_cmangle}"
 		else
-			return "{mmodule.c_name}__{mclass.c_name.to_cmangle}"
+			return "{mmodule.c_name}_0_{mclass.c_name}"
```
Comment (Morriar) [inline src/model/model.nit]: Why 0 in the two last branches?

File: src/model/model.nit [hunk 4]
```diff
@@ -2018,42 +2035,66 @@ abstract class MPropDef
 	# Therefore the combination of identifiers is awful,
 	# the worst case being
 	#
-	# ~~~nitish
-	# "{mclassdef.mmodule.full_name}#{mclassdef.mclass.intro_mmodule.full_name}::{mclassdef.name}#{mproperty.intro_mclassdef.mmodule.full_name}::{mproperty.intro_mclassdef.name}::{name}"
-	# ~~~
+	#  * a property "p::m::A::x"
+	#  * redefined in a refinement of a class "q::n::B"
+	#  * in a module "r::o"
+	#  * so "r::o#q::n::B#p::m::A::x"
```
Comment (Morriar) [inline src/model/model.nit]: Can't wait for generic type disambiguation using this syntax!

File: src/model/model.nit [hunk 5]
```diff
@@ -2018,42 +2035,66 @@ abstract class MPropDef
...
+					res.append mproperty.intro_mclassdef.mmodule.name
+					res.append "::"
+				end
+			end
+			if mclassdef.mclass != mproperty.intro_mclassdef.mclass then
+				# precise "B" only if not the same class than "A"
+				res.append mproperty.intro_mclassdef.name
+				res.append "::"
+			end
+			# Always use "x"
```
Comment (Morriar) [inline src/model/model.nit]: Always use the "x" char? Or are you referring to the "x" property used in the example shown 40 lines before?

Comment (Morriar) [inline src/model/model.nit]: #nitpicking

File: src/model/model.nit [hunk 6]
```diff
@@ -2018,42 +2035,66 @@ abstract class MPropDef
...
-		res.append name
 		return res.to_s
 	end
 
 	redef var c_name is lazy do
 		var res = new FlatBuffer
 		res.append mclassdef.c_name
-		res.append "__"
-		if is_intro then
+		res.append "_1_"
```
Comment (Morriar) [inline src/model/model.nit]: Alright, I understand the 0 now.

**Comment by Morriar** [comment]
+1

**Comment by jpages** [comment]
+1

File: src/model/model.nit (hunk 4 above)
Comment (privat) [inline src/model/model.nit]: remplace 'x' par le nom du type virtuel.

**Comment by privat** [comment]
reroll avec prise en compte des commenaires

File: src/model/mmodule.nit [hunk 7]
```diff
@@ -104,6 +112,22 @@ class MModule
 		end
 	end
 
+	# The namespace used for entities according to their visibility `v`.
```
Comment (Morriar) [inline src/model/mmodule.nit]: "Namespace used for..."

File: src/model/mproject.nit [hunk 8]
```diff
@@ -27,6 +27,10 @@ class MProject
 	# The name of the project
 	redef var name: String
 
+	redef fun full_name: String do return name
```
Comment (Morriar) [inline src/model/mproject.nit]: No need for return type since it's a redef.

File: src/model/mproject.nit [hunk 9]
```diff
@@ -27,6 +27,10 @@ class MProject
...
+
+	redef var c_name: String = name.to_cmangle is lazy
```
Comment (Morriar) [inline src/model/mproject.nit]: Same here

**Comment by Morriar** [comment]
+1

**Pull request merged**